torch/ao/quantization/experimental/linear.py

```diff
@@ -0,0 +1,180 @@
+import torch
+import numpy as np
+import math
+
+from torch.nn.quantized.modules.utils import WeightedQuantizedModule
+from torch.ao.quantization.experimental.observer import APoTObserver
+from torch.ao.quantization.experimental.quantizer import quantize_APoT, dequantize_APoT
+
+class LinearAPoT(WeightedQuantizedModule):
+    r"""
+    A quantized linear module with quantized tensor as inputs and outputs
+    to support APoT quantization.
+    We adopt the same interface as `torch.nn.Linear`, see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to :class:`~torch.nn.Linear`, attributes will be randomly
+    initialized at module creation time and will be overwritten later
+
+    Attributes:
+        alpha: `alpha` qparam of output Quantized Tensor, type: Tensor
+        gamma: `gamma` qparam of output Quantized Tensor, type: Tensor
+        quantization_levels: `quantization_levels` qparam of output Quantized Tensor, type: Tensor
+        level_indices: `level_indices` qparam of output Quantized Tensor, type: Tensor
+    """
+
+    def __init__(self, weight2quantize: torch.Tensor, signed=False):
+        # self.in_features = in_features
+        # self.out_features = out_features
+
+        super().__init__()
+
+        # hard code to APoT paper example inputs: b=4, k=2
+        observer = APoTObserver(b=4, k=2)
+
+        observer(weight2quantize)
+
+        self.alpha, self.gamma, self.quantization_levels, self.level_indices = observer.calculate_qparams(signed=signed)
+
+        quantized = quantize_APoT(weight2quantize, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
+        self.weight = dequantize_APoT(quantized)
+        self.weight = self.weight.reshape(weight2quantize.shape)
+
+    def decompose_APoT(self):
+        r"""
+        Helper function to decompose sums of PoT from APoT quantization
+        into corresponding tuples of PoT terms.
+        """
+        size = int(math.sqrt(torch.numel(self.weight)))
+
+        # decompose APoT weight
+        p_all = []
+
+        # create the "levels"
+        p0 = torch.tensor([0, 2**0, 2**-2, 2**-4])
```

Comment:
its weird that this is hardcoded, makes a lot more sense to set B, K as object vars and use them throughout the derive such values

Reply:
since i need to decompose the weight values it was easier to write it out explicitly like this -- since b and k are hardcoded (like how scale and zero_point are hardcoded in [uniform linear](https://www.internalfb.com/code/fbsource/[adba684270a7242e9856163a61480777f88fd57f]/fbcode/caffe2/torch/nn/quantized/modules/linear.py?lines=237) i thought this should be ok